test/Npgsql.Tests/PoolTests.cs

```diff
@@ -138,6 +138,36 @@ public async Task TimeoutGettingConnectorFromExhaustedPoolAsync()
             using (var conn3 = new NpgsqlConnection(connString))
                 conn3.Open();
         }
+        
+        [Test]
+        public void OverflowExceptionWhenTooManyWaiting()
+        {
+            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) {
+                ApplicationName = nameof(OverflowExceptionWhenTooManyWaiting),
+                MaxPoolSize = 1,
+            }.ToString();
+
+            using (var conn = new NpgsqlConnection(connString))
+            {
+                conn.Open();
+                Assert.True(PoolManager.TryGetValue(connString, out var pool));
+                var state = pool!.State;
+
+                try
+                {
+                    var newState = state;
+                    newState.Waiting = int.MaxValue;
+                    pool!.State = newState;
+                    var conn2 = new NpgsqlConnection(connString);
+                    Assert.Catch<OverflowException>(() => conn2.Open());
+                }
+                finally
+                {
+                    // Restore state for the closes work correctly.
+                    pool!.State = state; 
```

Comment:
Out of curiosity, what was the behavior without the `finally`? Was the dispose path throwing? Or is this just here to cleanup the bad pool?

Reply:
Cleanup the bad pool. If you don't the pool implementation tries to hand off a connection to a waiting tcs that isn't there and it thinks it raced the enqueue (as the change is waiting++ but also enqueue) so it loops forever.